src/ucp/stream/stream_recv.c

```diff
@@ -30,110 +62,311 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
 
     UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
 
-    if (ucs_queue_is_empty(&ep->ext.stream->data)) {
+    rdesc = ucp_stream_recv_data_nb_internal(ep->ext.stream);
+    if (ucs_likely(rdesc != NULL)) {
+        *length = rdesc->length;
+        ret = ucp_stream_rdesc_data(rdesc);
+    } else {
         ret = UCS_STATUS_PTR(UCS_OK);
-        goto out;
     }
 
-    rdesc = ucs_queue_pull_elem_non_empty(&ep->ext.stream->data, ucp_recv_desc_t,
-                                          stream_queue);
-    *length = rdesc->length;
-    ret = ucp_stream_rdesc_data(rdesc);
-
-out:
     UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
 
     return ret;
 }
 
+static UCS_F_ALWAYS_INLINE void
+ucp_stream_rdesc_release(ucp_recv_desc_t *rdesc)
+{
+    if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
+        uct_iface_release_desc(UCS_PTR_BYTE_OFFSET(rdesc,
+                                                   -sizeof(ucp_eager_sync_hdr_t)));
+    } else {
+        ucs_mpool_put_inline(rdesc);
+    }
+}
+
 UCS_PROFILE_FUNC_VOID(ucp_stream_data_release, (ep, data),
                       ucp_ep_h ep, void *data)
 {
     ucp_recv_desc_t *rdesc = ucp_stream_rdesc_from_data(data);
 
     UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
 
-    if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
-        uct_iface_release_desc((char*)rdesc - sizeof(ucp_eager_sync_hdr_t));
-    } else {
-        ucs_mpool_put_inline(rdesc);
-    }
+    ucp_stream_rdesc_release(rdesc);
 
     UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
 }
 
+static UCS_F_ALWAYS_INLINE ssize_t
+ucp_stream_rdata_unpack(void *rdata, size_t length, ucp_request_t *dst_req)
+{
+    /* Truncated error is not actual for stream, need to adjust */
+    size_t       valid_len = ucs_min((dst_req->recv.length -
+                                      dst_req->recv.state.offset),
+                                     length);
+    ucs_status_t status    = ucp_dt_unpack(dst_req->recv.datatype,
```

Comment:
since valid_len is used here, i would put these as regular code, not as part of variable initialization